tests/phpunit/tests/rest-api/rest-request.php

```diff
@@ -650,4 +650,57 @@ public function test_set_param_follows_parameter_order() {
 			$request->get_json_params()
 		);
 	}
+
+	public function test_set_param_updates_param_in_json_and_query() {
+		$request = new WP_REST_Request();
+		$request->add_header( 'content-type', 'application/json' );
+		$request->set_method( 'POST' );
+		$request->set_body(
+			wp_json_encode(
+				array(
+					'param' => 'value_body',
+				)
+			)
+		);
+		$request->set_query_params(
+			array(
+				'param' => 'value_query',
+			)
+		);
+		$request->set_param( 'param', 'new_value' );
+
+		$this->assertEquals( 'new_value', $request->get_param( 'param' ) );
+		$this->assertEquals( array( 'param' => 'new_value' ), $request->get_json_params() );
+		$this->assertEquals( array( 'param' => 'new_value' ), $request->get_query_params() );
+	}
+
+	public function test_set_param_updates_param_if_already_exists_in_query() {
+		$request = new WP_REST_Request();
+		$request->add_header( 'content-type', 'application/json' );
+		$request->set_method( 'POST' );
+		$request->set_body(
+			wp_json_encode(
+				array(
+					'param_body' => 'value_body',
+				)
+			)
+		);
+		$original_defaults = array(
+			'param_query' => 'default_query_value',
+			'param_body'  => 'default_body_value',
+		);
+		$request->set_default_params( $original_defaults );
+		$request->set_query_params(
+			array(
+				'param_query' => 'value_query',
+			)
+		);
+		$request->set_param( 'param_query', 'new_value' );
+
+		$this->assertEquals( 'new_value', $request->get_param( 'param_query' ) );
+		$this->assertEquals( array( 'param_body' => 'value_body' ), $request->get_json_params() );
+		$this->assertEquals( array( 'param_query' => 'new_value' ), $request->get_query_params() );
+		// Verify the default wasn't overwritten.
```

Comment:
Let's also check for `get_body_params`.

Reply:
Hmmmm this brings up an interesting case: we never called `set_body_params()`, so `get_body_params` return an empty array because `$this->params['POST']` was an empty array.

And because `set_param` is only setting the value on params that had it set previously (which it wasn't), `$this->params['POST']['param_query']` never gets set.

Is that weird?

Reply:
> Is that weird?

I think that behavior makes sense.